implementation/src/main/java/io/smallrye/config/EnvConfigSource.java

```diff
@@ -177,4 +142,141 @@ Object readResolve() {
             return new EnvConfigSource();
         }
     }
+
+    static final class EnvProperty {
+        private final String name;
+
+        public EnvProperty(final String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final EnvProperty that = (EnvProperty) o;
+            return equals(this.name, that.name);
+        }
+
+        @Override
+        public int hashCode() {
+            int h = 0;
+            int length = name.length();
+            if (length >= 2) {
+                if (name.charAt(length - 1) == '_' && name.charAt(length - 2) == '_') {
+                    length = length - 1;
+                }
+            }
+
+            for (int i = 0; i < length; i++) {
+                char c = name.charAt(i);
+                if (i == 0) {
+                    if (c == '%' || c == '_') {
+                        h = 31 * h + 31;
+                        continue;
+                    }
+                }
+
+                if (c == '.' || c == '_' || c == '-' || c == '"' || c == '*' || c == '[' || c == ']' || c == '/') {
+                    continue;
+                }
+                h = 31 * h + Character.toLowerCase(c);
+            }
+            return h;
+        }
+
+        static boolean equals(final String name, final String other) {
+            if (Objects.equals(name, other)) {
```

Comment:
inline what `equals` will do here: the "type profile" (ie the concrete type seen in the `Object::equals` method) of this method risk to have seen different other concrete types already, and adding `String` too, will made the performed `String::equals` within to become megamorphic.
This is called type profile pollution: see https://wiki.openjdk.org/display/HotSpot/MethodData

Given that `String` override `equals` and `Objects::equals`  is calling `Object:equals` on the not-null parameter, there's a chance to slow down due to megamoprhic virtual calls there

Reply:
Actually we only need the `==` and we can skip `String.equals`, because the rest of the method can handle that and we avoid 2 passes.